### PR TITLE
Implemented `messaging_servicebus` crate

### PR DESCRIPTION
**Pull request by danbugs**
This PR is addressing and hoping to close issue #255 — That is, I've updated the crate to use the new Azure core's `HttpClient`, as requested. Plus, coming from MindFlavor's AzureSDKForRust, I had to make some changes to error handling, so I did that too following the pattern used by other crates under sdk/.

Signed-off-by: Dan Chiarlone <dchiarlone@microsoft.com>

### CODE REVIEW

File: sdk/messaging_servicebus/Cargo.toml
Comment (cataggar) [inline sdk/messaging_servicebus/Cargo.toml:18]: Why?

**Review by cataggar — CHANGES_REQUESTED** [review]
Thanks for this. Sorry, we moved the goal post. Instead of `thiserror`, can you please use `azure_core::error`. See #771.

File: sdk/messaging_servicebus/Cargo.toml
Comment (danbugs) [reply sdk/messaging_servicebus/Cargo.toml:18]: I was using this crate in a local project to test it and, with `azure_core = { path = "../core", version = "0.2" }`, the compiler screamed at me that I was using two different `azure_core` versions (i.e., locally I had `azure_core = "0.2"`). After I changed it like so, it all worked 🤷

**Comment by danbugs** [comment]
@cataggar ~ Thank you for the review! Like I mentioned in an earlier comment, I was testing this project locally, and, while doing so, I realized none of the functions related to getting messages (i.e., `receive_and_delete`, `peek_lock`, or `peek_lock_full`) work — no wonder there's no test coverage for anything but `send_event` lol

Looking further into it, I realized that's because, rather than asking for a simple queue, we are sneakily asking for an event hub, which doesn't really have that sort of functionality as can be seen per our REST API reference:
image
> Note: [source](https://docs.microsoft.com/en-us/rest/api/eventhub/)

Considering event hubs are built on top of service buses, the user could just pass in a queue name instead of an event hub name and that would work fine-ish, but I figured that's just bad UX for a crate called messaging_**servicebus**.

I have an upcoming PR where I'm re-organizing things:
- renamed stuff (e.g., `event_hub` to `queue`).
- because we don't need different functions for every single type of request we want to prepare (e.g., `send_event_prepare`, and so on), so I'm unifying all those into a single `prepare_request` function — similar to how other crates under sdk/ do it.
- considering `delete_message`, `renew_lock`, and `unlock_message` only work when we have a `lock-id`, which is only ever returned in the `PrepareLockResponse` object, I made them member functions to that struct instead.
- added documentation comments and dev comments.
- removed the `duration` parameter that was passed to pretty much every function of our crate and hid that detail from the user — mostly because I think it's confusing to have to pass a SAS duration when I want to, for example, just get a message. I just made it a set (and safe) expiry duration of 1 hour.
- updated the existing example (we might want to add more later).
- added test coverage for everything. 

Also, per your request, I'll drop the `thiserror` dependency and review the `errors` module (some stuff there might not be being used).

Considering this was a _lot_ of work, what are your thoughts on making this a `0.3`?

**Comment by danbugs** [comment]
Another thing I'd like to point out is that, as soon as I push these latest changes, the tests from `messaging_servicebus` will probably fail because I changed one of the required environment variables from `AZURE_EVENT_HUB_NAME` to `AZURE_QUEUE_NAME`.

File: sdk/messaging_servicebus/Cargo.toml
Comment (rylev) [reply sdk/messaging_servicebus/Cargo.toml:18]: This should be changed back. If you also want to use `azure_core` as a direct dependency locally, you should write `azure_core = { path = "../YOUR/AZURE_CORE/CHECKOUT", version = "0.2" }` .

File: sdk/messaging_servicebus/src/lib.rs
Comment (rylev) [inline sdk/messaging_servicebus/src/lib.rs:4]: This line can be removed `extern crate` declarations are not necessary post Rust 2015

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (rylev) [inline sdk/messaging_servicebus/src/service_bus/client.rs:17]: This is fine, but do note that we will want to move this crate to using the pipeline architecture (as can be seen [here](https://github.com/Azure/azure-sdk-for-rust/blob/a76ebf194eddf0a1846ffec77078359a5735f5a4/sdk/data_cosmos/src/clients/cosmos.rs#L69-L76)) ASAP. However, we don't need to block this PR on that.

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (rylev) [inline sdk/messaging_servicebus/src/service_bus/client.rs:78]: We should move this error handling to a common method so that we don't have to repeat the error handling in every method. 

Also nit: error messages should all start with lowercase letters.

File: sdk/messaging_servicebus/src/service_bus/mod.rs
Comment (rylev) [inline sdk/messaging_servicebus/src/service_bus/mod.rs:26]: I don't think this should be inlined (it's called pretty rarely). If the compiler thinks it better to inline, we can let it do so.

File: sdk/messaging_servicebus/src/service_bus/mod.rs
Comment (rylev) [inline sdk/messaging_servicebus/src/service_bus/mod.rs:64]: Same as above

File: sdk/messaging_servicebus/tests/service_bus.rs
Comment (rylev) [inline sdk/messaging_servicebus/tests/service_bus.rs:17]: Does this expect `send_message_test` to run first? it seems like we want these tests to be one test instead of many.

File: sdk/messaging_servicebus/Cargo.toml
Comment (danbugs) [reply sdk/messaging_servicebus/Cargo.toml:18]: I don't want to use `azure_core` as a local dependency on my project, I want to use `azure_core = "0.2"`. With that said, why do I get errors for using two different `azure_core` versions if both `messaging_servicebus` and my local project use version `0.2`? That is, regardless of it being local or not...? 

For reference, this is the error I get:
```rs
error[E0308]: mismatched types
  --> crates/mq-azure-servicebus/src/lib.rs:33:13
   |
33 |             http_client,
   |             ^^^^^^^^^^^ expected trait `azure_core::http_client::HttpClient`, found trait `HttpClient`
   |
   = note: expected struct `Arc<(dyn azure_core::http_client::HttpClient + 'static)>`
              found struct `Arc<dyn HttpClient>`
   = note: perhaps two different versions of crate `azure_core` are being used?

For more information about this error, try `rustc --explain E0308
```

This is w/ my local project depending on `azure_core = "=0.2"`, and `messaging_servicebus` depending on `azure_core =  { path = "../core", version = "=0.2" }`. If I change `messaging_servicebus` to depend on `azure_core = "0.2"`, the problem is fixed. 

P.S. I'm getting the `HttpClient` the exact same way I do in the tests: `let http_client = azure_core::new_http_client();
`

Comment (danbugs) [reply sdk/messaging_servicebus/Cargo.toml]: In any case~ done ✅

File: sdk/messaging_servicebus/src/lib.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/lib.rs]: done ✅

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/client.rs:17]: Sounds good~ I'll leave this for now, so we can at least get a working `messaging_servicebus` crate out!

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/client.rs]: done ✅

File: sdk/messaging_servicebus/src/service_bus/mod.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/mod.rs]: done ✅

File: sdk/messaging_servicebus/tests/service_bus.rs
Comment (danbugs) [reply sdk/messaging_servicebus/tests/service_bus.rs:17]: That's how I had done it initially, I'll revert back to that ~ done ✅

File: sdk/messaging_servicebus/src/service_bus/mod.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/mod.rs]: done ✅

File: sdk/messaging_servicebus/Cargo.toml
Comment (cataggar) [inline sdk/messaging_servicebus/Cargo.toml:3]: You can revert this back to `0.2.0`. There is only a `0.1.0` published. https://crates.io/crates/azure_messaging_servicebus/versions

Comment (danbugs) [reply sdk/messaging_servicebus/Cargo.toml]: done ✅

File: sdk/messaging_servicebus/Cargo.toml
Comment (cataggar) [inline sdk/messaging_servicebus/Cargo.toml:28]: Please remove `thiserror`.

File: sdk/messaging_servicebus/src/utils.rs
Comment (cataggar) [inline sdk/messaging_servicebus/src/utils.rs:28]: I haven't looked at this library enough to know if a conversion to a `String` for a body is the right thing or if it should be passed around as `Bytes`.

Comment (danbugs) [reply sdk/messaging_servicebus/src/utils.rs:28]: I followed what the crate initially did there. Plus, as a user of the crate, I'd want `String` rather than having to add a dependency for `Bytes` and handle that.

File: sdk/messaging_servicebus/Cargo.toml
Comment (danbugs) [reply sdk/messaging_servicebus/Cargo.toml]: done ✅

File: sdk/messaging_servicebus/Cargo.toml
Comment (rylev) [reply sdk/messaging_servicebus/Cargo.toml]: The note `perhaps two different versions of crate `azure_core` are being used?` points to exactly what is happening.

You are using two versions of `azure_core` even though they are both labeled as version 0.2. One is the local copy (from the local `messaging_servicebus`) and one is from crates.io (from your direct dependency). The correct way to fix this is for your project to also depend on the local copy of `azure_core`.

Comment (danbugs) [reply sdk/messaging_servicebus/Cargo.toml]: @rylev ~ I feel like that can't be the case. I, for example, use the `storage_blobs` crate just fine without having to depend on a local `azure_core`, and that crate itself depends on a local `azure_core` dependency...

Comment (danbugs) [reply sdk/messaging_servicebus/Cargo.toml]: You're right~ I didn't know Rust would get confused by using two crates of equal versions from different registries. Thanks, Ryan!

**Comment by danbugs** [comment]
@rylev, and @cataggar ~ Is this good to merge? Once we do, we'll have to trigger the new release of the crate!

File: sdk/messaging_servicebus/src/utils.rs
Comment (cataggar) [reply sdk/messaging_servicebus/src/utils.rs:28]: We are okay with returning `Bytes`. A recent similar change I made is [here](https://github.com/Azure/azure-sdk-for-rust/pull/776/files#diff-5e3e9406014c7c90e2c7dd3f7616bc517d15bfdc63b60e47b55eb7bdd7d2a024L66). I'm not going to hold up this PR for this.

File: sdk/messaging_servicebus/examples/service_bus00.rs
Comment (Mossaka) [inline sdk/messaging_servicebus/examples/service_bus00.rs:30]: it would be great to have 
1. multi-thread example using rust channels, where one thread sends msg to the queue, and another thread recieves the msg from the queue. 
2. multi sender-receiver where one sender sends msg to multiple queues, and mutiple recivers are able to subscribe to different queues. 

what do you think?

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (Mossaka) [inline sdk/messaging_servicebus/src/service_bus/client.rs:51]: no need to make self mutable reference. 
```suggestion
    pub async fn send_message(&self, msg: &str) -> Result<(), Error> {
```

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (Mossaka) [inline sdk/messaging_servicebus/src/service_bus/client.rs:64]:
```suggestion
    pub async fn receive_and_delete_message(&self) -> Result<String, Error> {
```

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (Mossaka) [inline sdk/messaging_servicebus/src/service_bus/client.rs:87]:
```suggestion
        &self,
```

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (Mossaka) [inline sdk/messaging_servicebus/src/service_bus/client.rs:109]:
```suggestion
        &self,
```

**Review by Mossaka — CHANGES_REQUESTED** [review]

File: sdk/messaging_servicebus/examples/service_bus00.rs
Comment (danbugs) [reply sdk/messaging_servicebus/examples/service_bus00.rs:30]: Hmmm, I don't know if we should block the PR on examples.

Also, I should note: the Azure Service Bus API does not include the concept of "subscribing".

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/client.rs:109]: done ✅

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/client.rs:87]: done ✅

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/client.rs:51]: done ✅

File: sdk/messaging_servicebus/src/service_bus/client.rs
Comment (danbugs) [reply sdk/messaging_servicebus/src/service_bus/client.rs:16]: done ✅

**Pull request merged**
